remote_attestation_sessions/src/client.rs

```diff
@@ -0,0 +1,101 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use crate::SessionId;
+use anyhow::Context;
+use async_trait::async_trait;
+use oak_remote_attestation::handshaker::{
+    AttestationBehavior, ClientHandshaker, Encryptor, ServerIdentityVerifier,
+};
+
+#[async_trait]
+pub trait UnaryClient {
+    async fn message(&mut self, session_id: SessionId, body: Vec<u8>) -> anyhow::Result<Vec<u8>>;
+}
+
+/// gRPC Attestation Service client implementation.
+pub struct GenericAttestationClient<T: UnaryClient> {
+    session_id: SessionId,
+    encryptor: Encryptor,
+    client: T,
+}
+
+impl<T: UnaryClient> GenericAttestationClient<T> {
+    pub async fn create(
+        mut client: T,
+        expected_tee_measurement: &[u8],
+        server_verifier: ServerIdentityVerifier,
+    ) -> anyhow::Result<Self> {
+        let session_id: SessionId = rand::random();
+
+        let mut handshaker = ClientHandshaker::new(
+            AttestationBehavior::create_peer_attestation(expected_tee_measurement),
+            server_verifier,
+        );
+        let client_hello = handshaker
+            .create_client_hello()
+            .context("Couldn't create client hello message")?;
+
+        let mut response = client
+            .message(session_id, client_hello)
+            .await
+            .context("Couldn't message client hello message")?;
+
+        while !handshaker.is_completed() {
+            let request = handshaker
+                .next_step(&response)
+                .context("Couldn't process handshake message")?;
+
+            if let Some(request) = request {
+                response = client
+                    .message(session_id, request)
+                    .await
+                    .context("Couldn't message client hello message")?;
+            }
+        }
+
+        let encryptor = handshaker
+            .get_encryptor()
+            .context("Couldn't get encryptor")?;
+
+        Ok(Self {
+            session_id,
+            encryptor,
+            client,
+        })
+    }
+
+    /// Sends data encrypted by the [`Encryptor`] to the server and decrypts the server responses.
+    pub async fn message(&mut self, payload: Vec<u8>) -> anyhow::Result<Option<Vec<u8>>> {
+        let encrypted_request = self
+            .encryptor
+            .encrypt(&payload)
+            .context("Couldn't encrypt request")?;
+
+        let encrypted_response = self
+            .client
+            .message(self.session_id, encrypted_request)
+            .await
+            .context("Couldn't message encrypted data request")?;
+
+        let encoded_response = self
```

Comment:
The structure of this method is nicely symmetric, but the naming seems inconsistent (e.g. payload vs encoded_response). How about request_plaintext_bytes and response_plaintext_bytes or something like that?

Reply:
Agreed that'd be a nicer name, updated. Should note that this naming wasn't changed, this is just a moved line, L106 from grpc_unary_attestation/src/client.rs

Reply:
Actually ended up reverting to `encoded_response` after spending some more time in the remote attestation code. 

Yes, doesn't seem clear in the isolated context here. but pretty sure it was named this way as the bytes are an encoded proto that is decoded in `oak_functions/client/rust/src/lib.rs` L59.

The serialized proto is a bit counterintuitive, it's part of the items listed in: https://github.com/project-oak/oak/issues/2608

Reply:
I understand it's pre-existing, but it still seems not super intuitive to me; I would either rename, or add some comments around if possible.

Reply:
Happy to add a comment. :) And agreed it's confusing, imo the underlying issue is the asymmetry in what the runtime expects. It expects a raw byte slice in the request, but responds with an (encoded) proto that contains the reply and a status code. There's a bit of an inherent asymmetry.

Reply:
Added comments to explain :)

Reply:
Thanks, appreciated!